crates/ruff_server/src/session/settings.rs

```diff
@@ -1,25 +1,452 @@
-            .and_then(|code_action_settings| code_action_settings.resolve_support.as_ref())
-            .is_some_and(|resolve_support| resolve_support.properties.contains(&"edit".into()));
+use std::ops::Deref;
+
+use lsp_types::Url;
+use rustc_hash::FxHashMap;
+use serde::Deserialize;
+
+pub(crate) type WorkspaceSettingsMap = FxHashMap<Url, ClientSettings>;
+
+/// Built from the initialization options provided by the client.
+pub(crate) struct AllSettings {
+    pub(crate) global_settings: ClientSettings,
+    pub(crate) workspace_settings: WorkspaceSettingsMap,
+}
+
+/// Resolved client settings for a specific document. These settings are meant to be
+/// used directly by the server, and are *not* a 1:1 representation with how the client
+/// sends them.
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[allow(dead_code, clippy::struct_excessive_bools)]
+pub(crate) struct ResolvedClientSettings {
+    fix_all: bool,
+    organize_imports: bool,
+    lint_enable: bool,
+    disable_rule_comment_enable: bool,
+    fix_violation_enable: bool,
+}
+
+/// This is a direct representation of the settings schema sent by the client.
+#[derive(Debug, Deserialize, Default)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct ClientSettings {
+    fix_all: Option<bool>,
+    organize_imports: Option<bool>,
+    lint: Option<Lint>,
+    code_action: Option<CodeAction>,
+}
+
+/// This is a direct representation of the workspace settings schema,
+/// which inherits the schema of [`ClientSettings`] and adds extra fields
+/// to describe the workspace it applies to.
+#[derive(Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+struct WorkspaceSettings {
+    #[serde(flatten)]
+    settings: ClientSettings,
+    workspace: Url,
+}
+
+#[derive(Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+struct Lint {
+    enable: Option<bool>,
+}
+
+#[derive(Debug, Default, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+struct CodeAction {
+    disable_rule_comment: Option<CodeActionSettings>,
+    fix_violation: Option<CodeActionSettings>,
+}
+
+#[derive(Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(rename_all = "camelCase")]
+struct CodeActionSettings {
+    enable: Option<bool>,
+}
+
+/// This is the exact schema for initialization options sent in by the client
+/// during initialization.
+#[derive(Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(untagged)]
+enum InitializationOptions {
+    #[serde(rename_all = "camelCase")]
+    HasWorkspaces {
+        global_settings: ClientSettings,
+        #[serde(rename = "settings")]
+        workspace_settings: Vec<WorkspaceSettings>,
+    },
+    GlobalOnly {
+        settings: Option<ClientSettings>,
+    },
+}
+
+impl AllSettings {
```

Comment:
Nit: I don't know if this is unintentional or your persoanl preference. I prefer to have the impl block follow the `struct` definition because it allows me to read through the code from top to bottom without having to jump between struct definition and implementations and I think that's what we generally do in the Ruff code base

Reply:
I usually prefer to group struct definitions at the top and then put all the `impl`s below - I think it makes the struct definitions themselves easier to read, and I think it makes sense for the heavily-nested structs here. What I can do, though, is move `AllSettings` down to be closer to its associated `impl` block.